### PR TITLE
Allow credentials to be specified via environment variables

### LINKED ISSUE #2062: Support credential provided through environment variables
To provide initial support for Workbench, we have decided to support the specification of a server credential via environment variables. These are:

CONNECT_SERVER: the URL of the server
CONNECT_API_KEY: the ApliKey for the publisher
CONNECT_SERVER_NAME: <optional> the nickname of the server. If not specified, use the hostname of the server URL.

Do not allow the deletion of a credential specified via the environment variables.

Do not allow overlap of URL or Name with other credentials.

### PR DESCRIPTION
**Pull request by sagerb**
## Intent

Resolves #2062 
Partially addresses #1831 

This PR introduces automatic credentials, which are imported from environment variables.

## Type of Change

- - [ ] Bug Fix 
- - [x] New Feature 
- - [ ] Breaking Change 
- - [ ] Documentation 
- - [ ] Refactor 
- - [ ] Tooling 

## Approach

Since we do not want environment variables to persist, the actual keychain storage does not include these new types of credentials. Instead, they are inserted into the memory copy of the store and are removed before that store is persisted.

The environment credential has a fixed GUID of "00000000-0000-0000-0000-000000000000". This allows us to identify it, provide more contextual error messages, and restrict operations (deletes are not allowed).

Validation of the URL is not done at creation time. This will show up as a normal connectivity error within the pre-checks.

Conflict checks occur when credentials are created. They do not occur once credentials are in the store or environment.

I have selected different icons for the different types of credentials.

With the following entries within my environment (present before VSCode is launched - `~.zshenv`):

2024-08-01 at 1 25 PM

I now can see:

2024-08-01 at 1 23 PM

## Automated Tests

Go unit tests have been updated to cover new functionality.

## Directions for Reviewers

To validate this, you must set environment variables ahead of launching the first instance of VSCode. What this involved for me was to insert them into my `~.zshenv` file, close all instances of VSCode, and then restart VSCode, launching the extension directly or via debug.

Confirm that if no environment variables are set, then you do not see any extra credentials. 
- Validate that you can still add a new credential after this fix.
- Validate that credential creation will not allow overlap of the name or the URL.

Set the environment variables appropriately, and then relaunch VSCode. Validate that:
- you see the new credential
- you can use it to publish. 
- right-click does not show the delete option for the environment credential
- Validate that credential creation will not allow overlap of the Environment variable name or its URL.

## Checklist

- [ ] I have updated [CHANGELOG.md](../CHANGELOG.md) to cover notable changes.

### CODE REVIEW

File: internal/credentials/credentials.go [hunk 1]
```diff
@@ -71,9 +76,60 @@ func (cr *CredentialRecord) ToCredential() (*Credential, error) {
...
+// EnvFactory creates a Credential based on the presence of the
+// environment value of CONNECT_SERVER and CONNECT_API_KEY
+func (cs *CredentialsService) EnvCredentialRecordFactory() (*CredentialRecord, error) {
+
+	serverUrl := os.Getenv("CONNECT_SERVER")
+	ak := os.Getenv("CONNECT_API_KEY")
+	name := os.Getenv("CONNECT_SERVER_NAME")
+
+	if serverUrl != "" {
+		normalizedUrl, err := util.NormalizeServerURL(serverUrl)
```
[the remark this answers is not in the document]
Comment (jonkeane) [reply internal/credentials/credentials.go:88]: > I would even argue that while CONNECT_SERVER_NAME being optional is great, we may not even want to support it.

_nods_ yeah, is `CONNECT_SERVER_NAME` set on workbench instances? Or only `CONNECT_SERVER` `CONNECT_API_KEY`? If the name isn't set on workbench instances, agreed with Jordan we might not even support it right now

[the remark this answers is not in the document]
Comment (sagerb) [reply internal/credentials/credentials.go:88]: OK, removed.

File: internal/credentials/credentials.go [hunk 2]
```diff
@@ -71,9 +76,60 @@ func (cr *CredentialRecord) ToCredential() (*Credential, error) {
...
+				// if we can, we'll use the host for the name
+				name, _, err = net.SplitHostPort(u.Host)
+				if err != nil {
+					name = u.Host
+				}
+			}
+		}
+
+		cred := Credential{
+			GUID:   EnvVarGUID, // We'll use a well known GUID to indicate it is from the ENV vars
```
[the remark this answers is not in the document]
Comment (sagerb) [reply internal/credentials/credentials.go:106]: Thanks.

File: internal/credentials/credentials.go [hunk 3]
```diff
@@ -173,6 +255,14 @@ func (cs *CredentialsService) Set(name string, url string, ak string) (*Credenti
...
 // Saves the CredentialTable
 func (cs *CredentialsService) save(table CredentialTable) error {
+
+	// remove any environment variable credential from the table
+	// before saving
+	_, found := table[EnvVarGUID]
+	if found {
+		delete(table, EnvVarGUID)
+	}
+
```
[the remark this answers is not in the document]
Comment (sagerb) [reply internal/credentials/credentials.go:265]: I've wrapped the load and save methods to keep the logic separate.

File: internal/credentials/credentials.go [hunk 4]
```diff
@@ -192,9 +282,16 @@ func (cs *CredentialsService) load() (CredentialTable, error) {
 	data, err := ks.Get(ServiceName, "credentials")
 	if err != nil {
 		if err == keyring.ErrNotFound {
-			return make(map[string]CredentialRecord), nil
+			// create an empty table and continue (so we can add environment variables)
+			emptyTable := make(map[string]CredentialRecord)
+			emptyTableBytes, err := json.Marshal(emptyTable)
+			if err != nil {
+				return nil, fmt.Errorf("failed to initialize credentials: %v", err)
+			}
```
[the remark this answers is not in the document]
Comment (sagerb) [reply internal/credentials/credentials.go:290]: The code has been returned to the original save, so this is no longer applicable.

**Comment by sagerb** [comment]
PR requests have been committed in https://github.com/posit-dev/publisher/pull/2063/commits/dac31a82d083bd095f4dde0f91773e27bc9188f4

File: internal/credentials/credentials.go [hunk 5]
```diff
@@ -186,8 +281,27 @@ func (cs *CredentialsService) save(table CredentialTable) error {
...
+		return nil, err
+	}
+
+	// insert a possible environment variable credential before returning
+	record, err := cs.EnvCredentialRecordFactory()
+	if err != nil {
+		return nil, err
+	}
+	if record != nil {
+		table[EnvVarGUID] = *record
```
Comment (mmarchetti) [inline internal/credentials/credentials.go:297]: What should we do if there's a name or URL conflict here?

**Comment by mmarchetti** [comment]
The older infrastructure had explicit support for an `AccountList` that was populated by multiple providers, and we had providers for rsconnect, rsconnect-python, and environment variables. There is now a `CredentialsProvider` that reads from the keychain using the `CredentialsService`. Providers had an explicit source field so you could indicate where an account came from, without having to use a special GUID. This feature could be implemented at that level using the existing `envVarProvider`.  But, the credentials API reads directly from the `CredentialsService` and not from the `AccountList` so there would be work to do there. If we're not going to use the old Account structures, they can eventually be removed in favor of having the backend use Credentials everywhere instead of Accounts.

File: internal/credentials/credentials.go (hunk 5 above)
Comment (sagerb) [reply internal/credentials/credentials.go:315]: I've added error checks for the get, just to make sure the environment variable doesn't overlap with anything in the table.

**Comment by sagerb** [comment]
> The older infrastructure had explicit support for an `AccountList` that was populated by multiple providers, and we had providers for rsconnect, rsconnect-python, and environment variables. There is now a `CredentialsProvider` that reads from the keychain using the `CredentialsService`. Providers had an explicit source field so you could indicate where an account came from, without having to use a special GUID. This feature could be implemented at that level using the existing `envVarProvider`. But, the credentials API reads directly from the `CredentialsService` and not from the `AccountList` so there would be work to do there. If we're not going to use the old Account structures, they can eventually be removed in favor of having the backend use Credentials everywhere instead of Accounts.

When I started this, I wondered the same thing. Since we established that we were not going to support credentials coming in from reconnect and rsconnect-python, I figured it was not worth it to take the new implementation of credentials (vs accounts) back into that old infrastructure. We can certainly talk about it.

**Comment by mmarchetti** [comment]
We decided in the meeting to stick with this implementation, and eventually simplify or remove the `AccountList` infrastructure in the backend.

File: extensions/vscode/src/views/credentials.ts [hunk 6]
```diff
@@ -116,6 +116,7 @@ export class CredentialsTreeDataProvider
     }
     try {
       const api = await useApi();
+      item.cred.guid;
```
[the remark this answers is not in the document]
Comment (dotNomad) [reply extensions/vscode/src/views/credentials.ts:119]: Looks like this is still left in. We should remove this.

**Review by dotNomad — APPROVED** [review]
There is one line I think we should remove - it doesn't appear to do anything. I have a suggestion for that.

The separation of logic was fantastic. Made it much easier to reason about.

---

I did get into a bad state when I had both a keychain credential and new env var credential for Dogfood - which is expected we only want 1 credential per server. However the failure case was nothing in the extension could be loaded since credentials errored on request.

We may want a follow-up issue to handle that a bit better.

**Pull request merged**